### PR TITLE
[WIP] Contentious C-Chain Test

### PR DESCRIPTION
**Pull request by patrick-ogrady**
### TODO
* Fix finalized height check
* cleanup code
* remove tps log (won't be accurate)

### CODE REVIEW

File: testsuite/tests/cchain/bombard.go [hunk 1]
```diff
@@ -70,33 +78,78 @@ func (p *parallelBasicTxXputTest) ExecuteTest() error {
...
 		}
+		finalizedHeight = height
+		wg.Done()
+	}()
+
+	go func() {
+		for i := 0; i < p.numLists; i++ {
+			err := <-errs
+			if err != nil {
+				groupErr = err
```
Comment (aaronbuchwald) [inline testsuite/tests/cchain/bombard.go:118]: Could you log the individual errors here as they occur to ensure all of the errors are logged?

Comment (patrick-ogrady) [reply testsuite/tests/cchain/bombard.go:118]: Yeah this was just a rough initial spike...lots to clean up here

File: testsuite/tests/cchain/helpers.go [hunk 2]
```diff
@@ -80,3 +85,68 @@ func confirmTxList(ctx context.Context, client *ethclient.Client, txs []*types.T
...
+			b, err := c.BlockByNumber(ctx, big.NewInt(int64(i)))
+			if err != nil {
+				return err
+			}
+
+			blockHash := b.Hash().Hex()
+			if len(hash) == 0 {
+				hash = blockHash
+				continue
+			}
```
Comment (aaronbuchwald) [inline testsuite/tests/cchain/helpers.go:103]: Setting the hash to the blockHash looks a little suspicious to me, could we change this to get the block for the first client and then proceed to check that the rest are the same?

File: testsuite/tests/cchain/helpers.go [hunk 3]
```diff
@@ -80,3 +85,68 @@ func confirmTxList(ctx context.Context, client *ethclient.Client, txs []*types.T
...
+			consecutive++
+		} else {
+			consecutive = 0
+		}
+
+		if consecutive > consecutiveHeights {
+			return reportedHeight, nil
+		}
+
+		time.Sleep(100 * time.Millisecond)
```
Comment (aaronbuchwald) [inline testsuite/tests/cchain/helpers.go:150]: I think we should bump up either consecutiveHeights or the sleep statement to make sure it's really finalized

File: testsuite/tests/cchain/bombard.go [hunk 4]
```diff
@@ -35,12 +36,19 @@ func NewBasicTransactionThroughputTest(client *services.Client, numLists int, nu
 
 // ExecuteTest ...
 func (p *parallelBasicTxXputTest) ExecuteTest() error {
+	// create first client that funds rest of clients
+	// import funds to all addresses at start of test
+	funder := p.client[0]
```
Comment (patrick-ogrady) [inline testsuite/tests/cchain/bombard.go:41]: TODO: support passing in a single client to run a non-contentious test

File: testsuite/tests/cchain/helpers.go [hunk 5]
```diff
@@ -80,3 +85,68 @@ func confirmTxList(ctx context.Context, client *ethclient.Client, txs []*types.T
...
+func waitForStableTip(ctx context.Context, clients []*ethclient.Client) (uint64, error) {
+	var consecutive int
+
+	for {
+		var (
+			reportedHeight uint64
+			foundDiff      bool
+		)
+
+		for _, c := range clients {
```
Comment (patrick-ogrady) [inline testsuite/tests/cchain/helpers.go:123]: TODO: track if one client has been stuck on the same height for an extended period of time (behind others)

File: testsuite/tests/cchain/bombard.go [hunk 6]
```diff
@@ -70,33 +78,78 @@ func (p *parallelBasicTxXputTest) ExecuteTest() error {
...
 		}
 	}
+
 	finishedConfirming := time.Now()
-	logrus.Infof("Finished confirming after %v seconds. Total time start to finish: %v", finishedConfirming.Sub(finishedIssuing).Seconds(), finishedConfirming.Sub(launchedIssuers).Seconds())
+	logrus.Infof("Took %v to confirm txs", finishedConfirming.Sub(finishedVerifying))
+
+	totalTxs := p.numLists * p.numTxs
+	tps := float64(totalTxs) / stableTipDuration.Seconds()
+	logrus.Infof("Finalized %d transactions in %v (%f TPS). Checking time start to finish: %v", totalTxs, stableTipDuration, tps, finishedConfirming.Sub(launchedIssuers))
```
Comment (patrick-ogrady) [inline testsuite/tests/cchain/bombard.go:152]: Subtract time waiting at same height to confirm finalization from this time.

**Comment by patrick-ogrady** [comment]
Current issue:
```
Removed fairness-exceeding pending transaction hash
```

**Comment by patrick-ogrady** [comment]
Can get around this by broadcasting all as locals: https://github.com/ava-labs/coreth/blob/63aa0a10a15ff3dd25e34c60b97ac19fe64e5473/coreth.go#L118-L120

**Comment by patrick-ogrady** [comment]
### TODO
* transfer a large amount to one eth address
* distribute to each eth address will send from (much faster than cross-chain transfer for each)
* broadcast list of txs generated for all but funding address

**Comment by patrick-ogrady** [comment]
```
ERRO[0143] block production is stuck at [52 52 56 52 56]
INFO[0143] Took 2m14.976381785s to reach stable tip
ERRO[0143] Test failed due to block production is stuck at [52 52 56 52 56]
```

**Comment by patrick-ogrady** [comment]
Replaced by #41

**Pull request closed without merging**